### PR TITLE
Rename all getters to use get/get0 in name

### LINKED ISSUE #15236: Poorly named getter functions
There are a number of new functions that do not include a _get_: `XXX_provider()`, `XXX_number()`, `XXX_name()`, `XXX_properties()`, `XXX_description()` and likely more.  These are in EVP and ENCODER/DECODER namespaces.  The project is moving towards having _get_ in the name of getter functions.

Note: there is no corresponding setting for any of these, only the getters.

### PR DESCRIPTION
**Pull request by t8m**
For functions that exist in 1.1.1 provide a simple aliases via #define.

Fixes #15236

### CODE REVIEW

**Comment by t8m** [comment]
Please note this will not build as it changes just the headers to start the discussion. So please review the name changes.

I decided to rename everything even the functions that exist in 1.1.1 and provide aliases via macros otherwise the result would be totally inconsistent. Intentionally the old names are not deprecated because we do not want to make things harder for existing users just for the renames.

**Comment by richsalz** [comment]
Whew!  That's a lot of changes, and you haven't touched the docs yet :)  (Minor change will be needed to `man7/openssl-threads`)

**Comment by t8m** [comment]
Yes, the final PR will be quite huge. I hope it will be mostly a lot of sed invocations with some inevitable additional cleanup of indentation.

21. 5. 2021 17:53, 17:53, Rich Salz ***@***.***> napsal/a:

**Comment by t8m** [comment]
Another rebase done.

File: doc/man3/EVP_EncryptInit.pod [hunk 1]
```diff
@@ -489,19 +494,29 @@ Returns the B<EVP_CIPHER> structure when passed an B<EVP_CIPHER_CTX> structure.
...
 
-=item EVP_CIPHER_flags()
+=item EVP_CIPHER_get_flags()
 
 Returns any flags associated with the cipher. See EVP_CIPHER_meth_set_flags()
 for a list of currently defined flags.
 
+=item EVP_CIPHER_CTX_get_num() and EVP_CIPHER_CTX_set_num()
+
+Respectively gets or sets the cipher specific "num" parameter for the associated I<ctx>.
```
Comment (paulidale) [inline doc/man3/EVP_EncryptInit.pod:512]: Drop the respectively.  _Gets or sets the..._

File: doc/man3/EVP_EncryptInit.pod [hunk 2]
```diff
@@ -1019,27 +1034,38 @@ EVP_CIPHER_CTX_reset() returns 1 for success and 0 for failure.
...
 
-EVP_CIPHER_type() and EVP_CIPHER_CTX_type() return the NID of the cipher's
-OBJECT IDENTIFIER or NID_undef if it has no defined OBJECT IDENTIFIER.
+EVP_CIPHER_get_type() and EVP_CIPHER_CTX_get_type() return the NID of the
+cipher's OBJECT IDENTIFIER or NID_undef if it has no defined
+OBJECT IDENTIFIER.
 
 EVP_CIPHER_CTX_cipher() returns an B<EVP_CIPHER> structure.
 
+EVP_CIPHER_CTX_get_num() returns a nonnegative num value or
```
Comment (paulidale) [inline doc/man3/EVP_EncryptInit.pod:1059]: non-negative

File: doc/man3/EVP_EncryptInit.pod [hunk 3]
```diff
@@ -1019,27 +1034,38 @@ EVP_CIPHER_CTX_reset() returns 1 for success and 0 for failure.
...
+B<EVP_CTRL_RET_UNSUPPORTED> if the implementation does not support the call
+or on any other error.
+
+EVP_CIPHER_CTX_set_num() returns 1 on success and 0 if the implementation
+does not support the cal or on any other error.
```
Comment (paulidale) [inline doc/man3/EVP_EncryptInit.pod:1064]: _call_

File: include/openssl/evp.h [hunk 4]
```diff
@@ -601,19 +627,22 @@ OSSL_DEPRECATEDIN_3_0 unsigned char *EVP_CIPHER_CTX_iv_noconst(EVP_CIPHER_CTX *c
...
 void *EVP_CIPHER_CTX_get_cipher_data(const EVP_CIPHER_CTX *ctx);
 void *EVP_CIPHER_CTX_set_cipher_data(EVP_CIPHER_CTX *ctx, void *cipher_data);
-# define EVP_CIPHER_CTX_name(c)         EVP_CIPHER_name(EVP_CIPHER_CTX_get0_cipher(c))
-# define EVP_CIPHER_CTX_type(c)         EVP_CIPHER_type(EVP_CIPHER_CTX_get0_cipher(c))
+# define EVP_CIPHER_CTX_get0_name(c) EVP_CIPHER_get0_name(EVP_CIPHER_CTX_get0_cipher(c))
+# define EVP_CIPHER_CTX_get_type(c)  EVP_CIPHER_get_type(EVP_CIPHER_CTX_get0_cipher(c))
+# define EVP_CIPHER_CTX_type         EVP_CIPHER_CTX_get_type
 # ifndef OPENSSL_NO_DEPRECATED_1_1_0
-#  define EVP_CIPHER_CTX_flags(c)       EVP_CIPHER_flags(EVP_CIPHER_CTX_get0_cipher(c))
+#  define EVP_CIPHER_CTX_flags       EVP_CIPHER_get_flags
```
Comment (paulidale) [inline include/openssl/evp.h:642]: This one should still convert from the CIPHER_CTX to the CIPHER shouldn't it?

**Review by paulidale — COMMENTED** [review]
Look good.  A couple of documentation comments, one error and a suggestion..

**Comment by paulidale** [comment]
I would support merging this one quickly.  I imagine that the continuous updating to resolve conflicts will get old fast.

**Comment by paulidale** [comment]
Should we remove the various XXX_get_number() calls?  The integers return are internal and mostly meaningless for applications -- they aren't globally unique like NIDs and could even change between runs.

**Comment by t8m** [comment]
> Should we remove the various XXX_get_number() calls? The integers return are internal and mostly meaningless for applications -- they aren't globally unique like NIDs and could even change between runs.

Good point. Of course that should be a different PR. @mattcaswell @levitte Do you see any usefulness of these for 3rd party applications? I could imagine an application that might want to create some independent mapping table of the fetched algorithms and use the number for fast hash table indexing instead of using the name. But I actually do not know if there is any application that would want to do such thing. Perhaps some interactive crypto meta tool?

File: doc/man3/EVP_EncryptInit.pod (hunk 2 above)
Comment (t8m) [reply doc/man3/EVP_EncryptInit.pod:1059]: make doc-nits complains about non-negative.

File: test/defltfips_test.c [hunk 5]
```diff
@@ -36,7 +36,8 @@ static int test_is_fips_enabled(void)
     if (!TEST_ptr(sha256))
         return 0;
     if (is_fips
-            && !TEST_str_eq(OSSL_PROVIDER_name(EVP_MD_provider(sha256)), "fips")) {
+            && !TEST_str_eq(OSSL_PROVIDER_name(EVP_MD_get0_provider(sha256)),
```
[the remark this answers is not in the document]
Comment (t8m) [reply test/defltfips_test.c:39]: Yes! OSSL_PROVIDER_get0_name

**Comment by levitte** [comment]
> Should we remove the various XXX_get_number() calls? The integers return are internal and mostly meaningless for applications -- they aren't globally unique like NIDs and could even change between runs.

There are internal functions to fetch by number, so you shouldn't *remove* those functions, but I'm OK with making them internal as well.

Do note that those functions are documented to return a *dynamic* number, so why you think they should be constant or globally unique, I do not know.

File: doc/man3/EVP_DigestInit.pod [hunk 6]
```diff
@@ -12,16 +12,16 @@ EVP_MD_CTX_settable_params, EVP_MD_CTX_gettable_params,
...
-EVP_MD_CTX_type, EVP_MD_CTX_size, EVP_MD_CTX_block_size,
-EVP_MD_CTX_md_data, EVP_MD_CTX_update_fn, EVP_MD_CTX_set_update_fn,
+EVP_MD_CTX_get_type, EVP_MD_CTX_get_size, EVP_MD_CTX_get_block_size,
+EVP_MD_CTX_get0_md_data, EVP_MD_CTX_update_fn, EVP_MD_CTX_set_update_fn,
 EVP_md_null,
 EVP_get_digestbyname, EVP_get_digestbynid, EVP_get_digestbyobj,
-EVP_MD_CTX_pkey_ctx, EVP_MD_CTX_set_pkey_ctx,
+EVP_MD_CTX_get_pkey_ctx, EVP_MD_CTX_set_pkey_ctx,
 EVP_MD_do_all_provided
 - EVP digest routines
```
Comment (levitte) [inline doc/man3/EVP_DigestInit.pod:26]: No!  Don't *remove* the old names!  They're still documented, and most of all, someone doing `man EVP_MD_name` shouldn't suddenly get to know that there is no such man-page!

File: doc/man3/EVP_DigestInit.pod [hunk 7]
```diff
@@ -625,9 +625,9 @@ implementations of digests to be specified.
 If digest contexts are not cleaned up after use,
 memory leaks will occur.
 
-EVP_MD_CTX_name(), EVP_MD_CTX_size(), EVP_MD_CTX_block_size(),
-EVP_MD_CTX_type(), EVP_get_digestbynid() and EVP_get_digestbyobj() are defined
-as macros.
+EVP_MD_CTX_get0_name(), EVP_MD_CTX_get_size(), EVP_MD_CTX_get_block_size(),
+EVP_MD_CTX_get_type(), EVP_get_digestbynid() and EVP_get_digestbyobj() are
+defined as macros.
```
Comment (levitte) [inline doc/man3/EVP_DigestInit.pod:630]: Things went a little too automatically, eh?  Considering that you added documentation on the old names further down, you can remove this paragraph.

**Review by levitte — CHANGES_REQUESTED** [review]
I assume the rest of the documentation has had the same treatment, so what I say here should be applied on all changed pod files.

**Comment by paulidale** [comment]
> There are internal functions to fetch by number, so you shouldn't _remove_ those functions, but I'm OK with making them internal as well.

Let's do this.  In another PR.  These functions are effectively useless for an application.

File: util/libcrypto.num [hunk 8]
```diff
@@ -5043,14 +5043,14 @@ OSSL_PARAM_set_all_unmodified           5170	3_0_0	EXIST::FUNCTION:
...
+EVP_RAND_get0_name                      5175	3_0_0	EXIST::FUNCTION:
 EVP_RAND_is_a                           5176	3_0_0	EXIST::FUNCTION:
-EVP_RAND_provider                       5177	3_0_0	EXIST::FUNCTION:
+EVP_RAND_get0_provider                  5177	3_0_0	EXIST::FUNCTION:
 EVP_RAND_get_params                     5178	3_0_0	EXIST::FUNCTION:
 EVP_RAND_CTX_new                        5179	3_0_0	EXIST::FUNCTION:
 EVP_RAND_CTX_free                       5180	3_0_0	EXIST::FUNCTION:
-EVP_RAND_CTX_rand                       5181	3_0_0	EXIST::FUNCTION:
+EVP_RAND_CTX_get0_rand                  5181	3_0_0	EXIST::FUNCTION:
 EVP_RAND_CTX_get_params                 5182	3_0_0	EXIST::FUNCTION:
```
Comment (slontis) [inline util/libcrypto.num:5054]: Should the get_params & set_params calls be changed also?

Comment (paulidale) [reply util/libcrypto.num:5054]: Changed to what specifically?

Comment (slontis) [reply util/libcrypto.num:5054]: get0 ?

Comment (paulidale) [reply util/libcrypto.num:5054]: But these aren't doing a get0.  They copy the data into the params, so get1 is more appropriate.  Except for PTR types where there is no copy...

An argument could be made for the gettable/settable calls to include a 0 but that's a different matter.

Comment (t8m) [reply util/libcrypto.num:5054]: I do not think the get_/set_params should be renamed. The semantics is not a clear get1/get0. Anyway - a different PR if OTC decides to rename them.

Comment (slontis) [reply util/libcrypto.num:5054]: Not a problem - just thought I would mention it..

File: doc/man3/EVP_DigestInit.pod (hunk 7 above)
Comment (t8m) [reply doc/man3/EVP_DigestInit.pod:630]: But these calls are defined as macros. IMO this is correct and should be kept.

Comment (levitte) [reply doc/man3/EVP_DigestInit.pod:630]: 1.  No, EVP_MD_CTX_get0_name(), EVP_MD_CTX_get_size(), EVP_MD_CTX_get_block_size() and EVP_MD_CTX_get_type() are not defined as macros.
2.  You are saying *the same thing* in the paragraph that starts at line 721, i.e. this paragraph has become redundant.

Comment (t8m) [reply doc/man3/EVP_DigestInit.pod:645]: Richard, please look into the header. These calls _are_ defined as macros. And the paragraph in the HISTORY section is about the old names.

File: doc/man3/EVP_DigestInit.pod (hunk 6 above)
Comment (t8m) [reply doc/man3/EVP_DigestInit.pod:28]: Fixup pushed for this.

**Review by paulidale — APPROVED** [review]
LGTM

File: doc/man3/EVP_DigestInit.pod (hunk 7 above)
Comment (levitte) [reply doc/man3/EVP_DigestInit.pod:645]: [ahem]

Ok, blushing here...  I kept not seeing `CTX`.  Weird things that a brain can do at times.

So, [ahem] neeeeever mind!

**Comment by t8m** [comment]
Another rebase to fix conflicts. @paulidale Still OK? If you say so, I'll merge this.

**Comment by paulidale** [comment]
Go for it.

**Comment by t8m** [comment]
Aargh, another rebase needed. :100:

**Comment by t8m** [comment]
I'll merge after the checks pass.

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews!

**Pull request closed without merging**